web/src/main/java/com/microsoft/applicationinsights/web/internal/WebRequestTrackingFilterImpl.java

```diff
@@ -0,0 +1,204 @@
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.web.internal;
+
+import java.io.IOException;
+import java.util.Date;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.FilterChain;
+
+import com.microsoft.applicationinsights.TelemetryClient;
+import com.microsoft.applicationinsights.TelemetryConfiguration;
+import com.microsoft.applicationinsights.agent.internal.coresync.impl.AgentTLS;
+import com.microsoft.applicationinsights.internal.logger.InternalLogger;
+import com.microsoft.applicationinsights.internal.reflect.ClassDataUtils;
+
+/**
+ * Created by yonisha on 2/2/2015.
+ */
+public class WebRequestTrackingFilterImpl implements Filter {
+    // region Members
+
+    private WebModulesContainer webModulesContainer;
+    private boolean isInitialized = false;
+    private TelemetryClient telemetryClient;
+    private String key;
+    private boolean agentIsUp = false;
+
+    // endregion Members
+
+    // region Public
+
+
+    public void setKey(String key) {
+        try {
+            AgentTLS.getTLSKey();
+            agentIsUp = true;
+        } catch (Throwable throwable) {
+            agentIsUp = false;
+            InternalLogger.INSTANCE.logAlways(InternalLogger.LoggingLevel.ERROR, "setKey: Failed to find AgentTLS");
+        }
+        this.key = key;
+    }
+
+    public boolean isInitialized() {
+        return isInitialized;
+    }
+
+    /**
+     * Processing the given request and response.
+     * @param req The servlet request.
+     * @param res The servlet response.
+     * @param chain The filters chain
+     * @throws java.io.IOException Exception that can be thrown from invoking the filters chain.
+     * @throws javax.servlet.ServletException Exception that can be thrown from invoking the filters chain.
+     */
+    public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain) throws IOException, ServletException {
+        boolean isRequestProcessedSuccessfully = true;
+
+        setKeyOnTLS();
+
+        if (isInitialized) {
+            isRequestProcessedSuccessfully = invokeSafeOnBeginRequest(req, res);
+        }
+
+        try {
+            chain.doFilter(req, res);
+        } catch (ServletException se) {
```

Comment:
Instead of having 3 `catch`es here with the same code wouldn't it be nicer to have a single catch that will do the filtering, i.e.:

    catch (Exception e) {
        if (e instanceof ServletException || e instanceof IOException || e instanceof RuntimeException) {
            onException(e);
        }
        throw e;
    }